src/main/java/com/openshift/internal/restclient/okhttp/OpenShiftAuthenticator.java

```diff
@@ -0,0 +1,139 @@
+import com.openshift.restclient.authorization.IAuthorizationContext;
+import com.openshift.restclient.authorization.IAuthorizationDetails;
+import com.openshift.restclient.authorization.UnauthorizedException;
+import com.openshift.restclient.http.IHttpConstants;
+
+import okhttp3.Authenticator;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Request.Builder;
+import okhttp3.Response;
+import okhttp3.Route;
+
+/**
+ * OkHttp Authenticator implementations for OpenShift 3
+ * @author jeff.cantrill
+ *
+ */
+public class OpenShiftAuthenticator implements Authenticator, IHttpConstants{
+	
+	public static final String ACCESS_TOKEN = "access_token";
+	private static final String AUTH_ATTEMPTS = "X-OPENSHIFT-AUTH-ATTEMPTS";
+	private static final String ERROR = "error";
+	private static final String ERROR_DETAILS = "error_details";
+	
+	private Collection<IChallangeHandler> challangeHandlers = new ArrayList<>();
+	private OkHttpClient okClient;
+	private IClient client;
+	
+	@Override
+	public Request authenticate(Route route, Response response) throws IOException {
+		if(unauthorizedForCluster(response)){
+			String requestUrl = response.request().url().toString();
+			Request authRequest = new Request.Builder()
+					.addHeader("X-CSRF-Token", "1")
+					.url(route.address().url().toString() + "oauth/authorize?response_type=token&client_id=openshift-challenging-client")
+					.build();
+			try (
+				Response authResponse = tryAuth(authRequest)){
+				if(authResponse.isSuccessful()) {
+					String token = captureToken(authResponse);
+					return response.request().newBuilder()
+							.header(IHttpConstants.PROPERTY_AUTHORIZATION, String.format("%s %s", IHttpConstants.AUTHORIZATION_BEARER, token))
+							.build();
+				}
+			}
+			throw new UnauthorizedException(captureAuthDetails(requestUrl), ResponseCodeInterceptor.getStatus(response.body().string()));
+		}
+
+		return null;
+	}
+	
+	private boolean unauthorizedForCluster(Response response) {
+		String requestHost = response.request().url().host();
+		return response.code() == IHttpConstants.STATUS_UNAUTHORIZED && client.getBaseURL().getHost().equals(requestHost);
+	}
+	
+	private Response tryAuth(Request authRequest) throws IOException {
+		return okClient
+		.newBuilder()
+		.authenticator(new Authenticator() {
+			
+			@Override
+			public Request authenticate(Route route, Response response) throws IOException {
+				if(StringUtils.isNotBlank(response.request().header(AUTH_ATTEMPTS))) {
+					return null;
+				}
+				if(StringUtils.isNotBlank(response.header(IHttpConstants.PROPERTY_WWW_AUTHENTICATE))) {
+					for (IChallangeHandler challangeHandler : challangeHandlers) {
+						if(!challangeHandler.canHandle(response.headers())) {
+							Builder requestBuilder = response.request().newBuilder()
+									.header(AUTH_ATTEMPTS, "1");
+							return challangeHandler.handleChallange(requestBuilder).build();
+						}
+					}
+				}
+				return null;
```

Comment:
No, this is the indicator to stop calling authenticators per the lib documentation.